### PR TITLE
324: Add support to reject translations with feedback.

### LINKED ISSUE #324: Reject with feedback
From https://glotpress.trac.wordpress.org/ticket/335

> As an validator I want to decline a validation with a message so the translator know who declined it and why.
> This way I don't need to fix the translation and get's the original translator the credit it deserves.

### PR DESCRIPTION
**Pull request by superdav42**
Reviewers can now choose one or more predefined reasons when rejecting a translation and enter a feedback message to be shown to the translator. The data is stored in the gp_meta table.
Keyboard shortcuts also included.
Resolves #324 .

### CODE REVIEW

**Comment by Mte90** [comment]
Seems something that we have to integrate with https://github.com/GlotPress/GlotPress-WP/pull/804.
Can you add few screenshots?

**Comment by superdav42** [comment]
No problem.
screenshot from 2018-05-09 00-34-49
This appears after clicking the reject button.
It doesn't necessarily have to integrate with #804 but I could see value in storing the feedback in the notes table. I could change it pretty easily once it's merged or branch it from #804.

**Comment by Mte90** [comment]
The proposal of that pr was to create a custom table for feedback also to not have a big table with everything.
We will see after 3.0 what to do also with this pr.

**Comment by pixelverbieger** [comment]
Don't know what "branding" is supposed to be, but I recommend to add a reason like "wrong translation" or "bad translation" ;-)

File: gp-includes/routes/translation.php [hunk 1]
```diff
@@ -278,6 +279,12 @@ public function translations_post( $project_path, $locale_slug, $translation_set
 
 			if ( $this->can( 'approve', 'translation-set', $translation_set->id ) || $this->can( 'write', 'project', $project->id ) ) {
 				$set_status = 'current';
+				$user_id    = gp_post( 'user_id' );
```
Comment (akirk) [inline gp-includes/routes/translation.php:282]: This allows assigning any `user_id` to a translation. Wouldn't it be a more viable approach to add a flag that simply doesn't override the `user_id` further up in the code?

**Comment by akirk** [comment]
> The proposal of that pr was to create a custom table for feedback also to not have a big table with everything.

@mte90 I've been advising @superdav42 on his work here. At translate.wordpress.com we use the [gp-translation-helpers](https://github.com/Automattic/gp-translation-helpers) plugin to store comments as WordPress comments. It's debatable how much of a hack this is to be considered but I think this PR could (and maybe should) leave the commenting part aside and only implement reject reasons. Quoting from #324:

> As an validator I want to decline a validation with a message so the translator know who declined it and why.

Having a set of common reject reasons allows statistical evaluation and give basic feedback to the translator.

> This way I don't need to fix the translation and get's the original translator the credit it deserves.

This has been added with the latest commit. By first rejecting with a reason and then fixing it without stealing credit allows moving fast, not having to wait for a fix from the original translator who might still suggest a better translation after the fact.

I see the creation of an extra table for comments as @Mte90 implemented it in #804 as a more generic comments solution that would benefit from being scoped that way.

**Comment by toolstack** [comment]
I would think that reject reasons and comments are linked, in the sense that even with a reject reason you may need to add a comment to be more specific of what the error is.  Many translations are long and a simple reject reason may not give enough information to the translator to correct the issue.

**Comment by akirk** [comment]
From our experience though rejects are not the only reason for comments, especially if they can be made non-translation-set-specific. 

On May 17, 2018 11:17:36 AM UTC, Greg Ross <notifications@github.com> wrote:
>I would think that reject reasons and comments are linked, in the sense
>that even with a reject reason you may need to add a comment to be more
>specific of what the error is.  Many translations are long and a simple
>reject reason may not give enough information to the translator to
>correct the issue.
>
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/GlotPress/GlotPress-WP/pull/900#issuecomment-389832148

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by toolstack** [comment]
>From our experience though rejects are not the only reason for comments, especially if they can be made non-translation-set-specific.

Agreed, comments have a larger scope than rejects, but rejects needs comments.  I'd draw a venn diagram but my ASCII art skills aren't that good ;)

File: gp-includes/routes/translation.php (hunk 1 above)
Comment (superdav42) [reply gp-includes/routes/translation.php:282]: I started implementing this using a `keep_user` flag but it was tricky to retrieve the user_id for the current translation at this point in the code. If there is a function to do this I'd be happy to use it but I think the current implementation keeps things simple. The result of the action is creating a translation from POST values so passing the user_id seems appropriate. Only those with the 'approve' permission can do it so I don't think there's a security concern.

**Comment by superdav42** [comment]
Created a plugin version of this PR which projects can start using now if they need this feature. https://github.com/superdav42/gp-reject-feedback Should work with any versions since 2.0.0.

**Comment by Mte90** [comment]
Cool!
Sadly we have to wait the 3.0 release to start talking about this PR for 3.1.
Maybe if you join the glotpress channel on slack you can help us to move forward that release

**Comment by xavivars** [comment]
This would be extremely helpful. We often find extremely hard, when onboarding a new translator, to find a proper way to tell him/her why translations have been rejected.

File: gp-templates/translation-row.php [hunk 2]
```diff
@@ -143,6 +143,12 @@
 					<button class="ok" data-nonce="<?php echo esc_attr( wp_create_nonce( 'add-translation_' . $t->original_id ) ); ?>">
 						<?php echo $can_approve_translation ? __( 'Add translation &rarr;', 'glotpress' ) : __( 'Suggest new translation &rarr;', 'glotpress' ); ?>
 					</button>
+					<?php if ( $can_approve_translation ) : ?>
+						<?php _e( 'or', 'glotpress' ); ?>
+						<button class="ok" data-nonce="<?php echo esc_attr( wp_create_nonce( 'add-translation_' . $t->original_id ) ); ?>" data-user-id="<?php echo esc_attr( $t->user_id ); ?>" title="<?php echo esc_attr_e( 'Fixing a translation will keep credit for the original author.', 'glotpress' ); ?>">
+							<?php _e( 'Fix translation &rarr;', 'glotpress' ); ?>
+						</button>
+					<?php endif; ?>
```
Comment (ocean90) [inline gp-templates/translation-row.php:151]: This doesn't seem to be related to this PR (reject with feedback).

**Comment by amieiro** [comment]
This PR is outdated and its functionality has been added to GlotPress with the [gp-translation-helpers](https://github.com/GlotPress/gp-translation-helpers/) plugin, used at [translate.wordpress.org](https://translate.wordpress.org/). Can we close this PR?

**Comment by superdav42** [comment]
Yes
